interfaces/python_sdist/build_sdist.py

```diff
@@ -0,0 +1,199 @@
+    git_commit: str,
+    py_requires_ver_str: str,
+    cantera_version: str,
+    cantera_short_version: str,
+):
+    src_source = source_directory / "src"
+    src_target = target_directory / "src"
+    shutil.copytree(
+        src_source,
+        src_target,
+        ignore=shutil.ignore_patterns(
+            "fortran",
+            "clib",
+            "pch",
+            "global.cpp",
+            "SCons*",
+            "canteraStatic.cpp",
+        ),
+        dirs_exist_ok=True,
+    )
+    base_global_cpp = Path("base", "global.cpp")
+    replace_git_hash(
+        src_source / base_global_cpp, src_target / base_global_cpp, git_commit
+    )
+
+    include_source = source_directory / "include"
+    include_target = target_directory / "include"
+    # config.h needs to be filled on the user's machine by CMake. utils_utils.h is
+    # handled by the replace_git_hash command.
+    shutil.copytree(
+        include_source,
+        include_target,
+        ignore=shutil.ignore_patterns(
+            "clib",
+            "ext",
+            "config.h",
+            "utils_utils.h",
+        ),
+        dirs_exist_ok=True,
+    )
+    utils_utils_h = Path("cantera", "cython", "utils_utils.h")
+    replace_git_hash(
+        include_source / utils_utils_h,
+        include_target / utils_utils_h,
+        git_commit,
+    )
+
+    cantera_source = source_directory / "interfaces" / "cython" / "cantera"
+    cantera_target = target_directory / "cantera"
+    shutil.copytree(
+        cantera_source,
+        cantera_target,
+        ignore=shutil.ignore_patterns("__pycache__", "solution.py.in"),
+        dirs_exist_ok=True,
+    )
+
+    with_units = Path("with_units", "solution.py")
+    with_units_source = cantera_source / with_units.with_suffix(".py.in")
+    with_units_target = cantera_target / with_units
+    with_units_target.write_text(do_units_substitution(with_units_source))
+
+    data_source = source_directory / "data"
+    data_target = target_directory / "cantera" / "data"
+    shutil.copytree(
+        data_source,
+        data_target,
+        ignore=shutil.ignore_patterns("example_data"),
+        dirs_exist_ok=True,
+    )
+    shutil.copytree(
+        data_source / "example_data",
+        data_target / "example_data",
+        ignore=shutil.ignore_patterns(".git"),
+        dirs_exist_ok=True,
+    )
+
+    HERE = Path(__file__).parent
+    pyproject_toml_source = HERE / "pyproject.toml.in"
+    pyproject_toml_target = target_directory / "pyproject.toml"
+    pyproject_toml_target.write_text(
+        do_pyproject_substitution(
+            pyproject_toml_source, py_requires_ver_str, cantera_version
+        )
+    )
+
+    configure_source = HERE / "src" / "configure.py.in"
+    configure_target = target_directory / "src" / "configure.py"
+    configure_target.write_text(
+        do_configure_substitution(
+            configure_source, cantera_version, cantera_short_version
+        )
+    )
```

Comment:
That's a good point. This is actually a small problem... Since cmake will search the system for dependencies, I either need to pin CMake to only find the exact version we specify in the config.h for SUNDIALS and probably fmt, in which case this could be simplified; or, I need to allow those package versions to be configured at build time by CMake (which wouldn't be that hard), and then we need this double substitution since one needs to run at sdist build time and the other at wheel build time. Thoughts about which you'd prefer? I'm leaning towards the latter to support more downstream platforms.

Reply:
The only true variables in `config.h.in` are the cantera version and git commit. Or are you saying that you anticipate broadening the set of configured properties such that there would be other, system-dependent variables that needed to set in `config.h`, and `configure.py` would be responsible for doing that configuration?

Reply:
There's also at least the Sundials version, right? If that were to be something other than 7.0, it'd need to be configured on the build machine where the sdist is built to a wheel. That would have to be filled in by CMake

Reply:
But for now, you are just pinning the SUNDIALS version, right?

Reply:
This line: https://github.com/Cantera/cantera/pull/1727/files#diff-85b3f5eee3cabfe48b9a8d7164766ebe64a6bc3e408e62720bdc865dea854ed6R30 doesn't specify which version of SUNDIALS to look for on the system. So someone who builds the sdist locally and who has SUNDIALS installed on their system might end up with a version other than 7.0.

Reply:
Right, sorry... In CMake, the `FIND_APACKAGE_ARGS` sets `FetchContent` up to search the system first (using the standard `find_package` function) and only download the library with the given URL if it can't find it on the system. For the Linux builds, I created an image with our external dependencies already built, basically to save time compiling them on the ARM builds that are super slow. That's why it's picking up from the system on Linux. It's possible to set up `FIND_PACKAGE_ARGS` so that it only searches for a single version; it would also be possible to use the variable that CMake sets with the SUNDIALS version to write to the `config.h` at wheel build time.

On a related note, what's the best way for me to document how this is put together? I'm happy to add READMEs or pages in the docs or whatever works best to share this knowledge :smile:

Reply:
Ah, I see. I guess my inclination would be to keep this simple and only support single versions of these dependencies in the sdist. Hopefully, there shouldn't be very many users installing from the sdist itself, and if a user has more complicated requirements, the SCons build route is always there.

Regarding documentation, adding a new page on the sdist/wheel/PyPI builds in `doc/sphinx/develop/` would be helpful. I guess this sort of fits under the "how Cantera works" section on https://cantera.org/dev/develop/index.html.